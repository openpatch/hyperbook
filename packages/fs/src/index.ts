import fs from "fs/promises";
import path from "path";
import matter from "gray-matter";
import chalk from "chalk";
import {
  HyperbookFrontmatter,
  HyperbookJson,
  Hyperproject,
  HyperlibraryJson,
  Language,
  Link,
  Navigation,
  HyperbookPage,
  HyperbookSection,
  Glossary,
} from "@hyperbook/types";
import { findUp } from "find-up";

const listFiles = async (
  root: string,
  dir: string = "",
  absolute: boolean = true
): Promise<string[]> => {
  const files = await fs.readdir(path.join(root, dir));
  return Promise.all(
    files.flatMap(async (file) => {
      const stat = await fs.stat(path.join(root, dir, file));
      if (stat.isDirectory()) {
        return listFiles(root, path.join(dir, file), absolute);
      } else if (file.endsWith(".md")) {
        if (absolute) {
          return [path.join(root, dir, file)];
        } else {
          return [path.join(dir, file)];
        }
      }
      return [];
    })
  ).then((f) => f.flat());
};

export const readBook = async (root: string, absolute: boolean = true) => {
  return listFiles(path.join(root, "book"), "", absolute);
};

export const readGlossary = async (root: string, absolute: boolean = true) => {
  return listFiles(path.join(root, "glossary"), "", absolute);
};

export const makeGlossary = async (root: string): Promise<Glossary> => {
  const files = await readGlossary(root);
  const glossary: Glossary = {};

  for (const file of files) {
    const { content, data } = await readFile(file);
    let name = path.basename(file, ".md");
    if (data.name) {
      name = data.name;
    } else {
      console.log(
        `\n${chalk.yellow(
          `warn  `
        )}- Glossary page ${file} does not specify a name. Defaulting to the filename ${name}.`
      );
    }

    const letter = name[0].toUpperCase();
    if (!glossary[letter]) {
      glossary[letter] = [];
    }

    const relativePath = path
      .relative(root, file)
      .replace(/\.md$/, "")
      .split("/");

    glossary[letter].push({
      name,
      href: "/" + relativePath.join("/"),
    });
  }

  return glossary;
};

export const listPagesForTerm = async (
  root: string,
  term: string
): Promise<HyperbookPage[]> => {
  const files = await readBook(root);

  const pages: HyperbookPage[] = [];

  for (const file of files) {
    const { content, data } = await readFile(file);
    const r = new RegExp(`:t\\[.*\\]\\{#${term}(\..*)?\\}|:t\\[${term}\\]`);
    const m = content.match(r);

    if (m && !data.hide && data.name) {
      const relativePath = path
        .relative(root, file)
        .replace(/\.md$/, "")
        .split("/");
      const isIndex = relativePath[relativePath.length - 1] === "index";
      if (isIndex) {
        relativePath.pop();
      }
      pages.push({
        ...data,
        href: "/" + relativePath.join("/"),
      });
    }
  }

  return pages;
};

export const readFile = async (file: string) => {
  const source = await fs
    .readFile(file)
    .catch(() =>
      fs
        .readFile(file + ".md")
        .catch(() => fs.readFile(path.join(file, "index.md")))
    );

  const { content, data } = matter(source);

  return { content, data: data as HyperbookFrontmatter };
};

export const readHyperbook = async (root: string) => {
  return fs
    .readFile(path.join(root, "hyperbook.json"))
    .then((f) => f.toString())
    .then(JSON.parse) as Promise<HyperbookJson>;
};

export const readHyperlibrary = async (root: string) => {
  return fs
    .readFile(path.join(root, "hyperlibrary.json"))
    .then((f) => f.toString())
    .then(JSON.parse) as Promise<HyperlibraryJson>;
};

export const findHyperbook = async (file: string): Promise<HyperbookJson> => {
  return findUp("hyperbook.json", {
    cwd: file,
  } as any)
    .then((f) => {
      if (!f) {
        throw new Error("Could not find hyperbook.json");
      }
      return fs.readFile(f);
    })
    .then((f) => JSON.parse(f.toString()));
};

export const findHyperbookRoot = async (file: string): Promise<string> => {
  return findUp("hyperbook.json", {
    cwd: file,
  } as any).then((f) => {
    if (!f) {
      throw new Error("Could not find hyperbook.json");
    }
    return path.parse(f).dir;
  });
};

export const readProject = async (
  root: string,
  libraryEntry?: HyperlibraryJson["library"][0]
): Promise<Hyperproject> => {
  if (libraryEntry?.src) {
    root = path.join(root, libraryEntry.src);
  }
  const hyperbookJson = await readHyperbook(root).catch(() => null);
  if (hyperbookJson) {
    return {
      type: "book",
      src: root,
      basePath: libraryEntry?.basePath ?? hyperbookJson.basePath,
      name: libraryEntry?.name ?? hyperbookJson.name,
      icon: libraryEntry?.icon,
    };
  }

  const hyperlibraryJson = await readHyperlibrary(root).catch(() => null);
  if (hyperlibraryJson) {
    return {
      type: "library",
      src: root,
      basePath: libraryEntry?.basePath ?? hyperlibraryJson.basePath,
      name: libraryEntry?.name ?? hyperlibraryJson.name,
      projects: await Promise.all(
        hyperlibraryJson.library.map((p) =>
          readProject(root, {
            ...p,
            basePath: path.join(
              libraryEntry?.basePath ?? hyperlibraryJson.basePath ?? "",
              p.basePath
            ),
          })
        )
      ),
      icon: libraryEntry?.icon,
    };
  }

  console.log(
    `${chalk.red("Error")} - Missing book or library for path ${root}.`
  );

  throw Error(`Missing book or library for path ${root}`);
};

export const getProjectName = (project: Hyperproject, language?: Language) => {
  let label = "";
  if (typeof project.name === "string") {
    label = project.name;
  } else {
    if (language) {
      label = project.name[language] || "";
    } else {
      label = Object.values(project.name)[0];
    }
    if (!label) {
      console.log(
        chalk.red(
          `You need to provide a name for language ${language} in ${project.src}`
        )
      );
      throw Error("");
    }
  }
  return label;
};

type MakeLinkForHyperprojectOptions = {
  href?: {
    useSrc?: true;
    append?: string[];
    prepend?: string[];
    relative?: string;
    protocol?: string;
  };
};

export const makeLinkForHyperproject = async (
  project: Hyperproject,
  language: Language = "en",
  options: MakeLinkForHyperprojectOptions = {}
): Promise<Link> => {
  const label = getProjectName(project, language);

  if (project.type === "library") {
    return {
      label,
      links: await Promise.all(
        project.projects.map((p) =>
          makeLinkForHyperproject(p, language, options)
        )
      ),
      icon: project.icon,
    };
  } else {
    let href = project.basePath ?? "/";
    if (!href.startsWith("/")) {
      href = "/" + href;
    }

    if (options.href?.useSrc) {
      href = project.src;
    }
    if (options.href?.relative) {
      href = path.relative(options.href.relative, href);
    }
    if (options.href?.prepend) {
      href = path.join(...options.href.prepend, href);
    }
    if (options.href?.append) {
      href = path.join(href, ...options.href.append);
    }
    if (options.href?.protocol) {
      if (href.startsWith("/")) {
        href = href.slice(0, -1);
      }
      href = options.href.protocol + href;
    }

    if (href.length > 1 && href.endsWith("/")) {
      href = href.slice(0, -1);
    }

    return {
      label,
      href,
      icon: project.icon,
    };
  }
};

export const makeNavigationForHyperbook = async (
  root: string,
  currPath: string = "/"
): Promise<Navigation> => {
  const hyperbook = await readHyperbook(root);

  const getSectionsAndPages = async function (
    dirPath: string,
    pageList: HyperbookPage[] = []
  ) {
    const files = await fs.readdir(dirPath);
    let arrayOfPages: HyperbookPage[] = [];
    let arrayOfSections: HyperbookSection[] = [];

    for (const file of files) {
      let p = path.join(dirPath, file);
      let repo: string | null = null;
      if (hyperbook.repo) {
        repo = hyperbook.repo + "/" + path.relative(root, p);
      }

      const stat = await fs.stat(p);

      if (stat.isDirectory()) {
        const { pages, sections } = await getSectionsAndPages(p, pageList);
        const { content, data } = await readFile(p);
        const section: HyperbookSection = {
          ...data,
          href: "/" + path.relative(path.join(root, "book"), p),
          isEmpty: content.trim() === "",
          pages,
          sections,
        };
        if (repo) {
          section.repo = repo + "/index.md";
        }

        arrayOfSections.push(section);
      } else {
        const { data } = await readFile(p);
        if (p.endsWith(".md")) {
          p = p.substring(0, p.length - 3);
          if (path.relative(path.join(root, "book"), p) === "index") {
            p = p.substring(0, p.length - 5);
          }
          if (!p.endsWith("index")) {
            const page: HyperbookPage = {
              ...data,
              href: "/" + path.relative(path.join(root, "book"), p),
            };
            if (repo) {
              page.repo = repo;
            }

            arrayOfPages.push(page);
          }
        }
      }
    }

    arrayOfPages = arrayOfPages.sort((a, b) => (a.name > b.name ? 1 : -1));
    arrayOfPages = arrayOfPages.sort((a, b) => {
      const iIndex = a.index !== undefined ? a.index : 9999;
      const eIndex = b.index !== undefined ? b.index : 9999;
      return iIndex - eIndex;
    });
    arrayOfSections = arrayOfSections.sort((a, b) =>
      a.name > b.name ? 1 : -1
    );
    arrayOfSections = arrayOfSections.sort((a, b) => {
      const iIndex = a.index !== undefined ? a.index : 9999;
      const eIndex = b.index !== undefined ? b.index : 9999;
      return iIndex - eIndex;
    });

    return { pages: arrayOfPages, sections: arrayOfSections };
  };

  const getPageList = (
    sections: HyperbookSection[],
    pages: HyperbookPage[]
  ): HyperbookPage[] => {
    let pageList = [...pages];

    for (const section of sections) {
      pageList = [
        ...pageList,
        section,
        ...getPageList(section.sections, section.pages),
      ];
    }

    return pageList;
  };

  const { sections, pages } = await getSectionsAndPages(
    path.join(root, "book")
  );

  let pageList = getPageList(sections, pages);

  let i = pageList.findIndex((p) => p.href === currPath);
  const current = pageList[i] || null;

  pageList = pageList.filter((p) => !p.isEmpty && !p.hide);
  i = pageList.findIndex((p) => p.href === currPath);

  const next = pageList[i + 1] || null;
  const previous = pageList[i - 1] || null;

  return {
    next,
    current,
    previous,
    sections,
    pages,
  };
};
