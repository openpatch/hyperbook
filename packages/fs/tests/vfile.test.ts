import path from "path";
import { describe, it, expect } from "vitest";
import * as vfile from "../src/vfile";

describe("list", () => {
  const relative = (s: string) =>
    path.relative(path.join(__dirname, "fixtures"), s);
  let hyperbookPath = path.join(__dirname, "fixtures", "single-hyperbook");
  vfile.clean(hyperbookPath);

  const makeFileRelative = (f: vfile.VFile) => {
    if ("references" in f) {
      return {
        ...f,
        root: relative(f.root),
        path: { ...f.path, absolute: relative(f.path.absolute) },
        references: f.references.map((f) => ({
          ...f,
          root: relative(f.root),
          path: { ...f.path, absolute: relative(f.path.absolute) },
        })),
      } as vfile.VFileGlossary;
    }
    return {
      ...f,
      root: relative(f.root),
      path: { ...f.path, absolute: relative(f.path.absolute) },
    };
  };

  it("should list all", async () => {
    const files = await vfile.list(hyperbookPath);
    const relativeFiles = files.map(makeFileRelative);
    expect(relativeFiles).toMatchSnapshot();
  });

  it("should list for folder book", async () => {
    const files = await vfile.listForFolder(hyperbookPath, "book");
    expect(files.map(makeFileRelative)).toMatchSnapshot();
  });

  it("should include main index", async () => {
    const files = await vfile.listForFolder(hyperbookPath, "book");
    const file = files.find((f) => f.path.href === "/");
    expect(file).toBeDefined();
  });

  it("should get references", async () => {
    const files = await vfile.listForFolder(hyperbookPath, "glossary");
    expect(files.map(makeFileRelative)).toMatchSnapshot();
  });
});

describe("getMarkdown", () => {
  let hyperbookPath = path.join(__dirname, "fixtures", "single-hyperbook");
  vfile.clean(hyperbookPath);
  it("should get markdown from template", async () => {
    let files = await vfile.list(hyperbookPath);
    let templateFile = files.find(
      (f) => f.name === "use-template1" && f.folder === "book"
    );
    if (!templateFile) {
      throw Error("Template not found");
    }

    expect(await vfile.getMarkdown(templateFile)).toMatchSnapshot();
  });

  it("should get markdown from template for index", async () => {
    let files = await vfile.list(hyperbookPath);
    let templateFile = files.find(
      (f) => f.path.relative === "subsection2/index.yml" && f.folder === "book"
    );
    if (!templateFile) {
      throw Error("Template not found");
    }

    expect(await vfile.getMarkdown(templateFile)).toMatchSnapshot();
  });

  it("should get markdown for single use template", async () => {
    let files = await vfile.list(hyperbookPath);
    let templateFile = files.find(
      (f) =>
        f.path.relative === "subsection3/single-template.md.hbs" &&
        f.folder === "book"
    );
    if (!templateFile) {
      throw Error("Template not found");
    }

    expect(await vfile.getMarkdown(templateFile)).toMatchSnapshot();
  });
});

describe("getDirectory", () => {
  let hyperbookPath = path.join(__dirname, "fixtures", "single-hyperbook");
  vfile.clean(hyperbookPath);
  it("should include main index", async () => {
    let directory = await vfile.getDirectory(hyperbookPath, "book");
    expect(directory.index).toBeDefined();
  });
});

describe("extractLines", () => {
  const text =
    "Line 1\nLine 2\nLine 3\nLine 4\nLine 5\nLine 6\nLine 7\nLine 8\nLine 9\nLine 10";
  it("should extract lines by range", () => {
    const extractedLines = vfile.extractLines(text, "2-4", "...");
    expect(extractedLines).toMatchSnapshot();
  });
  it("should extract lines with gaps", () => {
    const extractedLines = vfile.extractLines(text, "1,4-6,8", "...");
    expect(extractedLines).toMatchSnapshot();
  });
  it("should extract lines with no ellipsis", () => {
    const extractedLines = vfile.extractLines(text, "1,4-6,8");
    expect(extractedLines).toMatchSnapshot();
  });
  it("should extract lines with custom ellipsis", () => {
    const extractedLines = vfile.extractLines(text, "1,4-6,8", "// ...");
    expect(extractedLines).toMatchSnapshot();
  });
  it("should extract lines and not add ellipsis on first and last line", () => {
    const extractedLines = vfile.extractLines(text, "1,4-6,10", "// ...");
    expect(extractedLines).toMatchSnapshot();
  });
  it("should return full content", () => {
    const extractedLines = vfile.extractLines(text);
    expect(extractedLines).toMatchSnapshot();
  });
  it("should extract lines matching a regex", () => {
    const source = `
import org.openpatch.scratch.Sprite;
import org.openpatch.scratch.Stage;
import org.openpatch.scratch.extensions.GifRecorder;

public class SpriteGetCurrentCostumeIndex {
    public SpriteGetCurrentCostumeIndex() {
        Stage myStage = new Stage(256, 100);
        Sprite mySprite = new Sprite("zeta", "examples/java/assets/zeta_green_badge.png");
        mySprite.addCostume("gamma", "examples/java/assets/gamma_purple_badge.png");
        mySprite.changeY(20);
        myStage.add(mySprite);
        mySprite.think("Index: " + mySprite.getCurrentCostumeIndex());
        GifRecorder recorder = new GifRecorder("examples/java/" + this.getClass().getName() + ".gif");
        recorder.start();

        myStage.wait(2000);
        mySprite.nextCostume();
        mySprite.think("Index: " + mySprite.getCurrentCostumeIndex());
        myStage.wait(2000);

        recorder.stop();
        System.exit(0);
    }

    public static void main(String[] args) {
        new SpriteGetCurrentCostumeIndex();
    }
}
`;
    const extracted = `
import org.openpatch.scratch.Sprite;
import org.openpatch.scratch.Stage;

public class SpriteGetCurrentCostumeIndex {
    public SpriteGetCurrentCostumeIndex() {
        Stage myStage = new Stage(256, 100);
        Sprite mySprite = new Sprite("zeta", "examples/java/assets/zeta_green_badge.png");
        mySprite.addCostume("gamma", "examples/java/assets/gamma_purple_badge.png");
        mySprite.changeY(20);
        myStage.add(mySprite);
        mySprite.think("Index: " + mySprite.getCurrentCostumeIndex());

        myStage.wait(2000);
        mySprite.nextCostume();
        mySprite.think("Index: " + mySprite.getCurrentCostumeIndex());
        myStage.wait(2000);

        System.exit(0);
    }

    public static void main(String[] args) {
        new SpriteGetCurrentCostumeIndex();
    }
}
`;
    const result = vfile.extractLines(source, "reg:[Rr]ecorder");
    expect(result).toEqual(extracted);
  });
});
