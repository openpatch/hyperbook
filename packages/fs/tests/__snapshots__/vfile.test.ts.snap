// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`extractLines > should extract lines and not add ellipsis on first and last line 1`] = `
"Line 1
// ...
Line 4
Line 5
Line 6
// ...
Line 10"
`;

exports[`extractLines > should extract lines by range 1`] = `
"...
Line 2
Line 3
Line 4
..."
`;

exports[`extractLines > should extract lines with custom ellipsis 1`] = `
"Line 1
// ...
Line 4
Line 5
Line 6
// ...
Line 8
// ..."
`;

exports[`extractLines > should extract lines with gaps 1`] = `
"Line 1
...
Line 4
Line 5
Line 6
...
Line 8
..."
`;

exports[`extractLines > should extract lines with no ellipsis 1`] = `
"Line 1
Line 4
Line 5
Line 6
Line 8"
`;

exports[`extractLines > should return full content 1`] = `
"Line 1
Line 2
Line 3
Line 4
Line 5
Line 6
Line 7
Line 8
Line 9
Line 10"
`;

exports[`getMarkdown > should get markdown for single use template 1`] = `
{
  "content": "# Single Use Template

TEST

Hi!
",
  "data": {},
}
`;

exports[`getMarkdown > should get markdown from template 1`] = `
{
  "content": "
Example use of a template.
",
  "data": {
    "name": "Use Template1 with YAML",
  },
}
`;

exports[`getMarkdown > should get markdown from template for index 1`] = `
{
  "content": "
Example use of a template.
",
  "data": {
    "name": "Use Template1 with YAML",
  },
}
`;

exports[`list > should get references 1`] = `
[
  {
    "extension": ".md",
    "folder": "glossary",
    "markdown": {
      "content": "",
      "data": {
        "name": "OOP",
      },
    },
    "name": "oop",
    "path": {
      "absolute": "single-hyperbook/glossary/oop.md",
      "directory": "",
      "href": "/glossary/oop",
      "relative": "oop.md",
    },
    "references": [
      {
        "extension": ".md",
        "folder": "book",
        "markdown": {
          "content": "
:t[OOP]{#oop}
",
          "data": {
            "name": "Paradigms",
          },
        },
        "name": "paradigms",
        "path": {
          "absolute": "single-hyperbook/book/paradigms.md",
          "directory": "",
          "href": "/paradigms",
          "relative": "paradigms.md",
        },
        "root": "single-hyperbook",
      },
    ],
    "root": "single-hyperbook",
  },
]
`;

exports[`list > should list all 1`] = `
[
  {
    "extension": ".md",
    "folder": "glossary",
    "markdown": {
      "content": "",
      "data": {
        "name": "OOP",
      },
    },
    "name": "oop",
    "path": {
      "absolute": "single-hyperbook/glossary/oop.md",
      "directory": "",
      "href": "/glossary/oop",
      "relative": "oop.md",
    },
    "references": [
      {
        "extension": ".md",
        "folder": "book",
        "markdown": {
          "content": "
:t[OOP]{#oop}
",
          "data": {
            "name": "Paradigms",
          },
        },
        "name": "paradigms",
        "path": {
          "absolute": "single-hyperbook/book/paradigms.md",
          "directory": "",
          "href": "/paradigms",
          "relative": "paradigms.md",
        },
        "root": "single-hyperbook",
      },
    ],
    "root": "single-hyperbook",
  },
  {
    "extension": ".md",
    "folder": "book",
    "markdown": {
      "content": "
:t[OOP]{#oop}
",
      "data": {
        "name": "Paradigms",
      },
    },
    "name": "paradigms",
    "path": {
      "absolute": "single-hyperbook/book/paradigms.md",
      "directory": "",
      "href": "/paradigms",
      "relative": "paradigms.md",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".md",
    "folder": "book",
    "markdown": {
      "content": "",
      "data": {
        "name": "Home",
      },
    },
    "name": "index",
    "path": {
      "absolute": "single-hyperbook/book/index.md",
      "directory": "",
      "href": "/",
      "relative": "index.md",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".json",
    "folder": "book",
    "markdown": {
      "content": "",
      "data": {},
    },
    "name": "empty",
    "path": {
      "absolute": "single-hyperbook/book/subsection/empty.json",
      "directory": "subsection",
      "href": "/subsection/empty",
      "relative": "subsection/empty.json",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".json",
    "folder": "book",
    "markdown": {
      "content": "",
      "data": {},
    },
    "name": "invalid",
    "path": {
      "absolute": "single-hyperbook/book/subsection/invalid.json",
      "directory": "subsection",
      "href": "/subsection/invalid",
      "relative": "subsection/invalid.json",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".yml",
    "folder": "book",
    "markdown": {
      "content": "
Example use of a template.
",
      "data": {
        "name": "Use Template1 with YAML",
      },
    },
    "name": "use-template1",
    "path": {
      "absolute": "single-hyperbook/book/subsection/use-template1.yml",
      "directory": "subsection",
      "href": "/subsection/use-template1",
      "relative": "subsection/use-template1.yml",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".md",
    "folder": "book",
    "markdown": {
      "content": "",
      "data": {
        "virtual": true,
      },
    },
    "name": "index",
    "path": {
      "absolute": "single-hyperbook/book/subsection/index.md",
      "directory": "subsection",
      "href": "/subsection",
      "relative": "subsection/index.md",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".md",
    "folder": "book",
    "markdown": {
      "content": "
# Test
",
      "data": {
        "name": "Test",
      },
    },
    "name": "test",
    "path": {
      "absolute": "single-hyperbook/book/subsection2/test.md",
      "directory": "subsection2",
      "href": "/subsection2/test",
      "relative": "subsection2/test.md",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".yml",
    "folder": "book",
    "markdown": {
      "content": "
Example use of a template.
",
      "data": {
        "name": "Use Template1 with YAML",
      },
    },
    "name": "index",
    "path": {
      "absolute": "single-hyperbook/book/subsection2/index.yml",
      "directory": "subsection2",
      "href": "/subsection2",
      "relative": "subsection2/index.yml",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".md.hbs",
    "folder": "book",
    "markdown": {
      "content": "# Single Use Template

TEST

Hi!
",
      "data": {},
    },
    "name": "single-template",
    "path": {
      "absolute": "single-hyperbook/book/subsection3/single-template.md.hbs",
      "directory": "subsection3",
      "href": "/subsection3/single-template",
      "relative": "subsection3/single-template.md.hbs",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".yml",
    "folder": "book",
    "markdown": {
      "content": "
Example use of a template.
",
      "data": {
        "name": "Use Template1 with YAML",
      },
    },
    "name": "index",
    "path": {
      "absolute": "single-hyperbook/book/subsection3/index.yml",
      "directory": "subsection3",
      "href": "/subsection3",
      "relative": "subsection3/index.yml",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".md",
    "folder": "book",
    "markdown": {
      "content": "",
      "data": {
        "name": "Test",
      },
    },
    "name": "test",
    "path": {
      "absolute": "single-hyperbook/book/subsection3/subsection31/test.md",
      "directory": "subsection3/subsection31",
      "href": "/subsection3/subsection31/test",
      "relative": "subsection3/subsection31/test.md",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".md",
    "folder": "book",
    "markdown": {
      "content": "
:t[OOP]{#oop}
",
      "data": {
        "name": "Paradigms",
      },
    },
    "name": "paradigms",
    "path": {
      "absolute": "single-hyperbook/book/paradigms.md",
      "directory": "",
      "href": "/paradigms",
      "relative": "paradigms.md",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".md",
    "folder": "book",
    "markdown": {
      "content": "",
      "data": {
        "name": "Home",
      },
    },
    "name": "index",
    "path": {
      "absolute": "single-hyperbook/book/index.md",
      "directory": "",
      "href": "/",
      "relative": "index.md",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".json",
    "folder": "book",
    "markdown": {
      "content": "",
      "data": {},
    },
    "name": "empty",
    "path": {
      "absolute": "single-hyperbook/book/subsection/empty.json",
      "directory": "subsection",
      "href": "/subsection/empty",
      "relative": "subsection/empty.json",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".json",
    "folder": "book",
    "markdown": {
      "content": "",
      "data": {},
    },
    "name": "invalid",
    "path": {
      "absolute": "single-hyperbook/book/subsection/invalid.json",
      "directory": "subsection",
      "href": "/subsection/invalid",
      "relative": "subsection/invalid.json",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".yml",
    "folder": "book",
    "markdown": {
      "content": "
Example use of a template.
",
      "data": {
        "name": "Use Template1 with YAML",
      },
    },
    "name": "use-template1",
    "path": {
      "absolute": "single-hyperbook/book/subsection/use-template1.yml",
      "directory": "subsection",
      "href": "/subsection/use-template1",
      "relative": "subsection/use-template1.yml",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".md",
    "folder": "book",
    "markdown": {
      "content": "",
      "data": {
        "virtual": true,
      },
    },
    "name": "index",
    "path": {
      "absolute": "single-hyperbook/book/subsection/index.md",
      "directory": "subsection",
      "href": "/subsection",
      "relative": "subsection/index.md",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".md",
    "folder": "book",
    "markdown": {
      "content": "
# Test
",
      "data": {
        "name": "Test",
      },
    },
    "name": "test",
    "path": {
      "absolute": "single-hyperbook/book/subsection2/test.md",
      "directory": "subsection2",
      "href": "/subsection2/test",
      "relative": "subsection2/test.md",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".yml",
    "folder": "book",
    "markdown": {
      "content": "
Example use of a template.
",
      "data": {
        "name": "Use Template1 with YAML",
      },
    },
    "name": "index",
    "path": {
      "absolute": "single-hyperbook/book/subsection2/index.yml",
      "directory": "subsection2",
      "href": "/subsection2",
      "relative": "subsection2/index.yml",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".md.hbs",
    "folder": "book",
    "markdown": {
      "content": "# Single Use Template

TEST

Hi!
",
      "data": {},
    },
    "name": "single-template",
    "path": {
      "absolute": "single-hyperbook/book/subsection3/single-template.md.hbs",
      "directory": "subsection3",
      "href": "/subsection3/single-template",
      "relative": "subsection3/single-template.md.hbs",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".yml",
    "folder": "book",
    "markdown": {
      "content": "
Example use of a template.
",
      "data": {
        "name": "Use Template1 with YAML",
      },
    },
    "name": "index",
    "path": {
      "absolute": "single-hyperbook/book/subsection3/index.yml",
      "directory": "subsection3",
      "href": "/subsection3",
      "relative": "subsection3/index.yml",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".md",
    "folder": "book",
    "markdown": {
      "content": "",
      "data": {
        "name": "Test",
      },
    },
    "name": "test",
    "path": {
      "absolute": "single-hyperbook/book/subsection3/subsection31/test.md",
      "directory": "subsection3/subsection31",
      "href": "/subsection3/subsection31/test",
      "relative": "subsection3/subsection31/test.md",
    },
    "root": "single-hyperbook",
  },
]
`;

exports[`list > should list for folder book 1`] = `
[
  {
    "extension": ".md",
    "folder": "book",
    "markdown": {
      "content": "
:t[OOP]{#oop}
",
      "data": {
        "name": "Paradigms",
      },
    },
    "name": "paradigms",
    "path": {
      "absolute": "single-hyperbook/book/paradigms.md",
      "directory": "",
      "href": "/paradigms",
      "relative": "paradigms.md",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".md",
    "folder": "book",
    "markdown": {
      "content": "",
      "data": {
        "name": "Home",
      },
    },
    "name": "index",
    "path": {
      "absolute": "single-hyperbook/book/index.md",
      "directory": "",
      "href": "/",
      "relative": "index.md",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".json",
    "folder": "book",
    "markdown": {
      "content": "",
      "data": {},
    },
    "name": "empty",
    "path": {
      "absolute": "single-hyperbook/book/subsection/empty.json",
      "directory": "subsection",
      "href": "/subsection/empty",
      "relative": "subsection/empty.json",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".json",
    "folder": "book",
    "markdown": {
      "content": "",
      "data": {},
    },
    "name": "invalid",
    "path": {
      "absolute": "single-hyperbook/book/subsection/invalid.json",
      "directory": "subsection",
      "href": "/subsection/invalid",
      "relative": "subsection/invalid.json",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".yml",
    "folder": "book",
    "markdown": {
      "content": "
Example use of a template.
",
      "data": {
        "name": "Use Template1 with YAML",
      },
    },
    "name": "use-template1",
    "path": {
      "absolute": "single-hyperbook/book/subsection/use-template1.yml",
      "directory": "subsection",
      "href": "/subsection/use-template1",
      "relative": "subsection/use-template1.yml",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".md",
    "folder": "book",
    "markdown": {
      "content": "",
      "data": {
        "virtual": true,
      },
    },
    "name": "index",
    "path": {
      "absolute": "single-hyperbook/book/subsection/index.md",
      "directory": "subsection",
      "href": "/subsection",
      "relative": "subsection/index.md",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".md",
    "folder": "book",
    "markdown": {
      "content": "
# Test
",
      "data": {
        "name": "Test",
      },
    },
    "name": "test",
    "path": {
      "absolute": "single-hyperbook/book/subsection2/test.md",
      "directory": "subsection2",
      "href": "/subsection2/test",
      "relative": "subsection2/test.md",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".yml",
    "folder": "book",
    "markdown": {
      "content": "
Example use of a template.
",
      "data": {
        "name": "Use Template1 with YAML",
      },
    },
    "name": "index",
    "path": {
      "absolute": "single-hyperbook/book/subsection2/index.yml",
      "directory": "subsection2",
      "href": "/subsection2",
      "relative": "subsection2/index.yml",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".md.hbs",
    "folder": "book",
    "markdown": {
      "content": "# Single Use Template

TEST

Hi!
",
      "data": {},
    },
    "name": "single-template",
    "path": {
      "absolute": "single-hyperbook/book/subsection3/single-template.md.hbs",
      "directory": "subsection3",
      "href": "/subsection3/single-template",
      "relative": "subsection3/single-template.md.hbs",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".yml",
    "folder": "book",
    "markdown": {
      "content": "
Example use of a template.
",
      "data": {
        "name": "Use Template1 with YAML",
      },
    },
    "name": "index",
    "path": {
      "absolute": "single-hyperbook/book/subsection3/index.yml",
      "directory": "subsection3",
      "href": "/subsection3",
      "relative": "subsection3/index.yml",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".md",
    "folder": "book",
    "markdown": {
      "content": "",
      "data": {
        "name": "Test",
      },
    },
    "name": "test",
    "path": {
      "absolute": "single-hyperbook/book/subsection3/subsection31/test.md",
      "directory": "subsection3/subsection31",
      "href": "/subsection3/subsection31/test",
      "relative": "subsection3/subsection31/test.md",
    },
    "root": "single-hyperbook",
  },
]
`;
